### PR TITLE
#47927: Export Categories CSV

### PR DESCRIPTION
**Pull request by Guccio163**
### Details

This PR focuses on adding export function to Workspace `Categories` section. Similar exports for `Tags` and `Members` will probably be incorporated into their import PRs after this PR gets approved.

User should have an option to click `Download CSV` button from three-dot option in the right-top corner. After that a file with categories in .csv format should be downloaded locally.

Screenshot 2024-09-06 at 11 57 10

### Fixed Issues

$ https://github.com/Expensify/App/issues/47927
PROPOSAL:

### Tests

[Non-android]

1. Navigate into workspace.
2. Enter `Categories` section.
3. Add some categories if you don'e have any.
4. Click the three-dot icon in top-right corner (`Import Spreadsheet` and `Download CSV` options should appear)
5. Select `Download CSV`.
6. "Categories-[time] [date]" file should be downloaded with all categories in .csv format.

[Android]

0. Check if there is 'Expensify' folder in 'Downloads' folder on the device.
1. Navigate into workspace.
2. Enter `Categories` section.
3. Add some categories if you don'e have any.
4. Click the three-dot icon in top-right corner (`Import Spreadsheet` and `Download CSV` options should appear)
5. Select `Download CSV`.
6. "Categories-[time] [date]" file should be downloaded with all categories in .csv format  ***in the Expensify folder***.
7. Repeat steps 2-5, the exported .csv file should be placed in the same Expensify folder as previous file.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

N/A

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/794dbb68-50f6-4f51-9cf8-1436f7265747

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/97fa4764-2340-4e54-a4ee-5368238f3482

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/18ed5c0e-8700-487e-82a7-a55253b55a74

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/2ed1e8b5-da6e-4183-8241-8d1f6219d0d3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/8726393e-54d9-4e04-bf28-135b09d97639

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/2a861e38-2342-4445-8723-1f2738ca20f6

</details>

### CODE REVIEW

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 1]
```diff
@@ -309,6 +309,17 @@ function WorkspaceCategoriesPage({route}: WorkspaceCategoriesPageProps) {
                     Navigation.navigate(ROUTES.WORKSPACE_CATEGORIES_IMPORT.getRoute(policyId));
                 },
             },
+            {
+                icon: Expensicons.Download,
+                text: translate('common.downloadCSV'),
```
Comment (filip-solecki) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:314]: add the translation key to `spreadsheet` instead of `common`

**Review by filip-solecki — APPROVED** [review]
LGTM, small adjustments to be done ✅

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 1 above)
Comment (Guccio163) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:314]: I was following "Import spreadsheet" example as `common.importSpreadsheet`; should I move it to spreadsheet too?

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 2]
```diff
@@ -309,6 +309,17 @@ function WorkspaceCategoriesPage({route}: WorkspaceCategoriesPageProps) {
...
+                onSelected: () => {
+                    if (isOffline) {
+                        setIsOfflineModalVisible(true);
```
[the remark this answers is not in the document]
Comment (Guccio163) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:317]: Great, thanks for catching this!

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 1 above)
Comment (filip-solecki) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:314]: Yes, do it please 👌

File: src/libs/actions/Policy/Category.ts [hunk 3]
```diff
@@ -793,6 +796,19 @@ function setPolicyDistanceRatesDefaultCategory(policyID: string, currentCustomUn
...
+    const finalParameters = enhanceParameters(WRITE_COMMANDS.EXPORT_CATEGORIES_CSV, {
+        policyID,
+    });
+
+    const formData = new FormData();
+    Object.entries(finalParameters).forEach(([key, value]) => {
+        formData.append(key, String(value));
+    });
+
+    fileDownload(ApiUtils.getCommandURL({command: WRITE_COMMANDS.EXPORT_CATEGORIES_CSV}), 'Categories.csv', '', false, formData, CONST.NETWORK.METHOD.POST);
```
[the remark this answers is not in the document]
Comment (dylanexpensify) [reply src/libs/actions/Policy/Category.ts:809]: @filip-solecki that sounds fine to me.

[the remark this answers is not in the document]
Comment (mountiny) [reply src/libs/actions/Policy/Category.ts:809]: Sounds good

**Comment by Guccio163** [comment]
Update: today I found a bug in exporting on Android; I've found a fix for it, but have to investigate further whether it will cause some problems. Have a great weekend!

**Review by rushatgabhane — APPROVED** [review]

File: src/libs/actions/Policy/Category.ts [hunk 4]
```diff
@@ -998,6 +1001,19 @@ function setPolicyDistanceRatesDefaultCategory(policyID: string, currentCustomUn
...
```
Comment (mountiny) [inline src/libs/actions/Policy/Category.ts:1014]: Can you make the `'Categories.csv'` a CONST?

**Review by mountiny — APPROVED** [review]
One NAB, but that can be addressed in a follow up

File: src/libs/actions/Policy/Category.ts (hunk 4 above)
Comment (Guccio163) [reply src/libs/actions/Policy/Category.ts:1014]: Sure, I'll try to remember to add it in `Members` or `Tags` import/export 👍

**Pull request merged**